Add libshaderc_spvc interface to SPIRV-Cross.

fjhenigman opened the pull request:
Not much here yet, though it has some tests and they pass.
At this point mostly want to see if the file paths and class names are agreeable.
The approach is shamelessly stolen from libshaderc.

fjhenigman commented:
@dneto0 ptal

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,149 @@
+#include <stddef.h>
+#include <stdint.h>
+
+//XXX share with libshaderc?
```
Yes, I think share this with libshaderc.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,149 @@
+//XXX share with libshaderc?
+// Indicate the status of a compilation.
+typedef enum {
+  spvc_compilation_status_success = 0,
```
I haven't reviewed the possible error states from SPIRV-Cross.  I presume this is representative.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,149 @@
+  spvc_compilation_status_internal_error,  // unexpected failure
+  spvc_compilation_status_null_result_object,
+  spvc_compilation_status_invalid_assembly,
+} spvc_compilation_status;
```
Regarding symbol names.  Should everything be spvc_* or shaderc_spvc_* ??'

I think spvc_ is reasonable.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,149 @@
+// Takes SPIR-V as a sequence of 32-bit words and compiles to GLSL.
+SHADERC_EXPORT spvc_compilation_result_t spvc_compile_into_glsl(
+    const spvc_compiler_t compiler, const uint32_t* source,
+    size_t source_len, const spvc_compile_options_t additional_options);
```
Is nullptr additional_options ok?

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,149 @@
+// function.
+typedef struct spvc_compilation_result* spvc_compilation_result_t;
+
+// Takes SPIR-V as a sequence of 32-bit words and compiles to GLSL.
```
Say that the SPIR-V must pass validation.
Which then begs the question of how to specify validation options.

One choice is to set, as an additional option, an spv_validator_options object as defined by spirv-tools/libspirv.h
(and then also include that header at the top of this file too.)

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.hpp:
```diff
@@ -0,0 +1,190 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SHADERC_SPVC_SPVC_HPP_
```
Include guard should match file path location.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.hpp:
```diff
@@ -0,0 +1,190 @@
+ private:
+  CompileOptions& operator=(const CompileOptions& other) = delete;
+  spvc_compile_options_t options_;
+  //std::unique_ptr<IncluderInterface> includer_;
```
Delete this.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.hpp:
```diff
@@ -0,0 +1,190 @@
+// Methods begin() and end() are also provided to enable range-based for.
+// They are synonyms to cbegin() and cend(), respectively.
+template <typename OutputElementType>
+class CompilationResult {
```
This template was needed for the old shaderc path because the output was either text or a sequence of 32-bit words.  But spirv-cross output is always a string.  So we don't need this complexity.

I'd add a TODO to simplify all this.

dneto0 commented on libshaderc_spvc/src/common_shaders_for_test.h:
```diff
@@ -0,0 +1,57 @@
+extern "C" {
+#endif
+
+const uint32_t kShader1[] = {
```
Please add a comment to say what original source was.  That will help us understand what it's testing.

dneto0 replied:
Also, later tests probably should use SPIR-V assembly to exercise wonky cases.  Although maybe the wonky case tests really belong in SPIRV-Cross.  spvc is about plumbing.

dneto0 commented on libshaderc_spvc/src/spvc.cc:
```diff
@@ -0,0 +1,106 @@
+#endif
+
+struct spvc_compile_options {
+  //spvc_target_env target_env = spvc_target_env_default;
```
Default target env should be Vulkan 1.0

dneto0 commented on README.md:
```diff
@@ -84,6 +84,7 @@ git clone https://github.com/google/googletest.git
 git clone https://github.com/google/glslang.git
 git clone https://github.com/KhronosGroup/SPIRV-Tools.git spirv-tools
 git clone https://github.com/KhronosGroup/SPIRV-Headers.git spirv-headers
+git clone https://github.com/KhronosGroup/SPIRV-Cross.git spirv-cross
```
I've had complaints that we place code in directories with names different from original repos.   So let's go with directory name SPIRV-Cross rather than spirv-cross.  Yes, that breaks with the pattern from spirv-tools and spirv-headers, but we're starting fresh with SPIRV-Cross here.

fjhenigman replied:
@Kangz isn't sure renaming is possible (something about case-insensitive filesystems - I don't see the problem myself) but said we could try.  Let me know if you want it and I'll do it as a separate set of changes to chrome, dawn, and shaderc.

fjhenigman commented:
Sorry I should have warned that was completely unpolished and not to look too closely.
Just wanted to get something running and sort out naming before doing too much.
I cleaned up everything @dneto0 mentioned and more.  Added a test for the C++ interface.
Still to do: BUILD.gn, Android.mk, the command line tool, and lots of API functionality.
@s-perron ptal.

s-perron commented on libshaderc_spvc/src/spvc.cc:
```diff
@@ -0,0 +1,148 @@
+
+void shaderc_spvc_compile_options_set_target_env(
+    shaderc_spvc_compile_options_t options, shaderc_target_env target,
+    uint32_t version) {
```
Would it be better to use the enum type `shaderc_env_version` instead of a uint32_t for the version?  The same would go for other places you use the version.

fjhenigman replied:
The other 'version' is for language version, and it looks like shaderc, glslang, spirv-cross all use a number for that rather than an enum.  I used uint32_t to match what spirv-cross uses.

dneto0 replied:
Yep.  It's traditional in this space to use a 32-bit int.  Even though that's begging for misinterpretation and name clashes.  At this point I go with the flow.

s-perron commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,117 @@
+// be issued.  The version will be for distinguishing between different versions
+// of the target environment.  The version value should be either 0 or
+// a value listed in shaderc_env_version.  The 0 value maps to Vulkan 1.0 if
+// |target| is Vulkan, and it maps to OpenGL 4.5 if |target| is OpenGL.
```
Why do we want to have a value for "0"?

fjhenigman replied:
see other comment about shaderc_spvc_compile_options_set_target_env()

fjhenigman commented on libshaderc_spvc/README.md:
```diff
@@ -0,0 +1,52 @@
+# libshaderc_spvc
+
+A library for validating and translating SPIR-V.
+
```
I think it's addressed.  I changed the README to say what @dneto0 wanted said, and the code does call the validator.
But I just noticed I don't bail out if validation fails, so I'll fix that.

s-perron replied:
Sorry something caused me to be looking at an older version.

fjhenigman commented:
I thought the PR was automatically tested, but I don't see that happening?
Maybe that's blocked by a reviewer saying "requested changes."
I'm not set up to test the Android build so I was hoping some bot would do that for me.

fjhenigman commented:
Thanks @dneto0 but waiting for others to add the label is not an ideal solution.  Can I get permission to do it myself?

dneto0 commented on libshaderc_spvc/src/spvc.cc:
```diff
@@ -0,0 +1,155 @@
+    spirv_cross::CompilerGLSL cross(source, source_len);
+    cross.set_common_options(options->glsl);
+    result->compiler_output = cross.compile();
+    // An exception during compiling would crash (if exceptions off) or jump to
```
Good comment!

dneto0 commented on libshaderc_spvc/src/spvc.cc:
```diff
@@ -0,0 +1,155 @@
+  std::string validation_messages;
+  std::string compiler_output;
+  shaderc_compilation_status status =
+      shaderc_compilation_status_null_result_object;
```
Arguably this could be .....compilation_error?
In the SPIRV-Cross crash case we would get a result object but this default status set. 

Anyway, I'm not to concerned about this.

fjhenigman replied:
But if it crashed there'd be no one around to check the status?
This is the same in libshaderc.

dneto0 commented on libshaderc_spvc/src/spvc.cc:
```diff
@@ -0,0 +1,155 @@
+    result->status = shaderc_compilation_status_success;
+  }
+  CATCH_IF_EXCEPTIONS_ENABLED(...) {
+    result->status = shaderc_compilation_status_compilation_error;
```
I think this should also set the compiler output string to an error message saying SPIRV-Cross failed.  Otherwise we have no clue about the nature of the failure.

dneto0 commented on libshaderc_spvc/src/spvc_c_smoke_test.c:
```diff
@@ -0,0 +1,82 @@
+    // Early exit on failure.
+    return -1;
+  }
+  shaderc_spvc_result_release(result);
```
In the success case, print something about the output, e.g. the length of the output string.

dneto0 commented on libshaderc_spvc/src/spvc_cpp_test.cc:
```diff
@@ -0,0 +1,35 @@
+
+  CompilationResult result = compiler.CompileSpvToGlsl(
+      kShader1, sizeof(kShader1) / sizeof(uint32_t), options);
+  ASSERT_EQ(shaderc_compilation_status_success, result.GetCompilationStatus());
```
Use the EXPECT form:  EXPECT_EQ

Also, EXPECT that the length of the resulting string is positive.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,114 @@
+    shaderc_spvc_compilation_result_t;
+
+// Takes SPIR-V as a sequence of 32-bit words and compiles to high level
+// language.
```
Say that the SPIR-V must pass validation.

dneto0 commented on libshaderc_spvc/include/shaderc/spvc.h:
```diff
@@ -0,0 +1,114 @@
+// Set language version.  Default is 450.
+SHADERC_EXPORT void shaderc_spvc_compile_options_set_language_version(
+    shaderc_spvc_compile_options_t options, uint32_t version);
+
```
Please add a TODO: add a method to set validator options.

Kangz commented on libshaderc_spvc/include/shaderc/spvc.hpp:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  // Which GLSL version should be produced.  Default is 450.
+  void SetLanguageVersion(uint32_t version) {
```
nit: SetOutputLanguageVersion. We'll also need a SetOutputLanguage to choose HLSL vs. GLSL etc.

Kangz commented:
C++ Interface looks usable by Dawn. My concern is that we'll need a lot more of it to be able to support a single backend. See what Dawn touches inside SPIRV-Cross in any of its backends: we need to set options to flipY and other things, rename bindings, gather combined samplers and textures on GL, etc.

dneto0 commented:
Also please rebase to pick up the DEPS changes
